pkg/build/util/util.go

```diff
@@ -307,3 +307,41 @@ func SafeForLoggingS2IConfig(config *s2iapi.Config) *s2iapi.Config {
 	newConfig.ScriptsURL, _ = s2iutil.SafeForLoggingURL(newConfig.ScriptsURL)
 	return &newConfig
 }
+
+// GetBuildEnv gets the build strategy environment
+func GetBuildEnv(build *buildapi.Build) *[]kapi.EnvVar {
+	switch {
+	case build.Spec.Strategy.SourceStrategy != nil:
+		return &build.Spec.Strategy.SourceStrategy.Env
+	case build.Spec.Strategy.DockerStrategy != nil:
+		return &build.Spec.Strategy.DockerStrategy.Env
+	case build.Spec.Strategy.CustomStrategy != nil:
+		return &build.Spec.Strategy.CustomStrategy.Env
+	case build.Spec.Strategy.JenkinsPipelineStrategy != nil:
+		return &build.Spec.Strategy.JenkinsPipelineStrategy.Env
+	default:
+		return nil
+	}
+}
+
+// UpdateBuildEnv updates the strategy environment
+// This will replace the existing variable definitions with provided env
+func UpdateBuildEnv(build *buildapi.Build, env []kapi.EnvVar) {
```

Comment:
Sorry, the concern I've got is if we allow substitution in the buildrequest envs.  Here's an example that is problematic:

existing strategy envs: A=1, B=2, C=3, D=$(B)
buildrequest envs: B=$(C)
resulting envs: A=1, B=$(C), C=3, D=$(B)
evaluates after substitution to: A=1, B=$(C), C=3, D=$(C)

I'm not quite sure what can actually be done about this.  Rejecting buildrequests with envs with substitutions might have been nice, but perhaps that ship has sailed?